### PR TITLE
babeld: Migrate from openwrt-routing

### PR DESCRIPTION
**Pull request by adrianschmutzler**
This migrates the babeld package from openwrt-routing together with the package-specific history.

---

**Migration tutorial:**

This is an example for migrating a package with its folder-based history from openwrt-routing.

This can be done very quickly (about two minutes) and with a few commands:

Clone the source repo
`git clone https://github.com/openwrt-routing/packages transfer`

Enter source repo
`cd transfer`

Create branch for changes ("extract")
`git checkout -b extract`

Extract commits for one directory (here babeld) from extract branch; this will move its contents to the repo root!
`git filter-branch --prune-empty --subdirectory-filter babeld extract`

Create temporary directory
`mkdir /data/tmp`

Move stuff to the desired subfolder (here net/babeld)
`git filter-branch -f --tree-filter 'mv * /data/tmp; mkdir -p net/babeld; mv /data/tmp/* net/babeld/'  --prune-empty extract`

Go to packages directory
`cd ../openwrt-packages`

Add remote for transfer project
`git remote add transfer ../transfer`

Pull in changes and rebase them to current master
```
git fetch transfer extract
git checkout transfer/extract
git rebase master
```

Now, add your changes and create PR
`git checkout -b addbabeld`
...

This PR is meant as a show-case on how packages could be transferred this way. If it is disliked, just close it again and move the package by different means.

### CODE REVIEW

**Comment by aparcar** [comment]
Are you sure you want all these commits per package? I honour you tutorial, just think this becomes messy...

**Comment by adrianschmutzler** [comment]
@aparcar I only know a small fraction of packages in openwrt-routing. From them, I have the feeling that commits affecting more than one package are rare and thus it would be still okay to duplicate them.

However, one could still just move the whole openwrt-routing master into the net subdir, without caring about individual packages, and then put this on top. This should be possible with a similar tutorial like mine here.

Still, with my approach one could keep together package history and the final changes required to merge them here.

Anyway, still was just meant as a show-case, so I won't be offended if you don't do it like this because it's not desirable.

**Comment by champtar** [comment]
@adrianschmutzler I would have imported all the commits & packages at once, I'll try to do an example today

**Comment by aparcar** [comment]
> @adrianschmutzler I would have imported all the commits & packages at once, I'll try to do an example today

News? I'd be interested to see how it's done.

Apart from that, isn't it bloat to add some 1420 commits to the master branch? Instead we could move the openwrt-routing/packages to openwrt/routing and mark it as archived. In parrallel a single (or many small) PR(s) transfer (and update) the routing.git packages?

**Comment by champtar** [comment]
@aparcar see #9419 
if we use the full history way we keep the original commit in this repo, so we can drop openwrt-routing entirely
we can then continue to use normal tools like git blame / git log to understand what happened in the recent past, and not switch back and forth between current and old repo

**Comment by adrianschmutzler** [comment]
Since this is for illustration only and not meant to be actually merged, I will close this PR.

**Comment by adrianschmutzler** [comment]
@kerneis @neheb @aparcar
I have reopened and added a commit on top of this which bring this (very simple) package in a state where it should be acceptable for merge.

I just need a hint on what I have to put for PKG_LICENSE_FILES.

buildpr fails due to the history, where several commits have mismatch of author and Signed-off, which I would ignore.

Some of the patches lack "babeld:" prefix, as they were treewide in openwrt-routing. I would personally just add the "babeld: " for those if agreed upon, which might be simple enough so one even does not have to add another Signed-off to those.

I intend to use this as an example case for whether this approach is acceptable. If yes, one could work on with other targets, and merge them here step-by-step.
If no, since adding full repository has received negative feedback, we at least have reached a decision and packages will be added without history ...

**Comment by adrianschmutzler** [comment]
I have also quickly done the same for batctl and batman-adv:
https://github.com/adrianschmutzler/packages/commits/addbatctl
https://github.com/adrianschmutzler/packages/commits/addbatmanadv

Since the approach is effectively doing a rebase, merge commits are lost, but history is fully supported in GitHub.

File: net/babeld/Makefile [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+PKG_NAME:=babeld
+PKG_VERSION:=1.8.4
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://www.irif.fr/~jch/software/files/
+PKG_HASH:=98070dc418c190f047b8d69eb47987df30ded8f0fca353c49427d3137ad08b87
+
+PKG_LICENSE:=MIT
+PKG_LICENSE_FILES:=
```
Comment (diizzyy) [inline net/babeld/Makefile]: Should be ```LICENCE```?
https://github.com/jech/babeld/blob/master/LICENCE

File: net/babeld/Makefile [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+# Copyright (C) 2007-2009 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=babeld
+PKG_VERSION:=1.8.4
```
Comment (diizzyy) [inline net/babeld/Makefile]: 1.8.5 is out

File: net/babeld/Makefile [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+endef
+
+define Package/babeld/conffiles
+/etc/babeld.conf
+/etc/config/babeld
+endef
+
+MAKE_FLAGS+= \
+	CFLAGS="$(TARGET_CFLAGS)" \
+	LDLIBS="" \
```
Comment (diizzyy) [inline net/babeld/Makefile]: Are these really needed?

**Comment by diizzyy** [comment]
https://github.com/jech/babeld/blob/master/Makefile#L4
As far as I can tell this overrides toolchain settings?

Needs be squashed into one commit

File: net/babeld/Makefile [hunk 4]
```diff
@@ -14,7 +14,12 @@ PKG_RELEASE:=1
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
 PKG_SOURCE_URL:=https://www.irif.fr/~jch/software/files/
 PKG_HASH:=98070dc418c190f047b8d69eb47987df30ded8f0fca353c49427d3137ad08b87
+
 PKG_LICENSE:=MIT
```
Comment (neheb) [inline net/babeld/Makefile]: PKG_MAINTAINER:= should go here. If none, leave it empty.

File: net/babeld/files/babeld.init [hunk 5]
```diff
@@ -197,6 +197,10 @@ start_service() {
 	procd_close_instance
```
Comment (neheb) [inline net/babeld/files/babeld.init]: There are minor warnings from shellcheck.

**Comment by adrianschmutzler** [comment]
Next weekend...

File: net/babeld/Makefile (hunk 2 above)
Comment (adrianschmutzler) [reply net/babeld/Makefile]: Update includes 1.9.1.

File: net/babeld/Makefile (hunk 1 above)
Comment (adrianschmutzler) [reply net/babeld/Makefile]: Updated.

File: net/babeld/Makefile (hunk 4 above)
Comment (adrianschmutzler) [reply net/babeld/Makefile]: Updated.

**Comment by adrianschmutzler** [comment]
I've pushed an updated version including latest patches from routing repo.

Unaddressed comments:
- MAKE_FLAGS   https://github.com/openwrt/packages/pull/9408#discussion_r309927681
- CDEBUGFLAGS   https://github.com/openwrt/packages/pull/9408#issuecomment-517487147
- babeld.init shellcheck   https://github.com/openwrt/packages/pull/9408#discussion_r309942056

Despite, I still do not see a _conclusion_ on whether this should be with full history or squashed into a single commit (I personally prefer the former).

@zorun @aparcar @dizzy @neheb

**Comment by aparcar** [comment]
Regarding full history or squashed, I'm in favor of the latter: Move routing.git to openwrt/routing.git and mark it as archived. Each package from routing.git could be added in a single clean commit, reducing confusion for others monitoring packages.git commits, not familiar with these new routing packages.

**Comment by adrianschmutzler** [comment]
BTW: Is there already any plan what happens to the stable branches (19.07, 18.06, 17.01)? Should the remain in routing (and be maintained there), or are they supposed to move, too ...

**Comment by neheb** [comment]
I say remain.

**Comment by adrianschmutzler** [comment]
Based on the feedback I received, the majority of responses favors merging without history. I will thus close this PR, as it was meant as proof-of-concept for the opposite and I do not want to merge this without support of the package maintainers.

**Comment by aparcar** [comment]
@adrianschmutzler thanks for working on this! I'll soon create some PRs as well.

**Pull request closed without merging**